sdks/python/apache_beam/ml/inference/base_test.py

```diff
@@ -171,6 +171,38 @@ def test_unexpected_inference_args_passed(self):
             FakeModelHandlerFailsOnInferenceArgs(),
             inference_args=inference_args)
 
+  def test_increment_failed_batches_counter(self):
+    with self.assertRaises(ValueError, FakeModelHandlerFailsOnInferenceArgs):
+      with TestPipeline() as pipeline:
+        examples = [1, 5, 3, 10]
+        pcoll = pipeline | 'start' >> beam.Create(examples)
+        inference_args = {'key': True}
+        _ = pcoll | base.RunInference(FakeModelHandlerFailsOnInferenceArgs(),
+                inference_args=inference_args)
+        run_result = pipeline.run()
+        run_result.wait_until_finish()
+
+        metric_results = (
+            run_result.metrics().query(MetricsFilter().with_name('failed_batches_counter')))
+        num_failed_batches_counter = metric_results['counters'][0]
+        self.assertEqual(num_failed_batches_counter.committed, 1)
```

Comment:
Is this guaranteed to be 1, or could it be larger if the example got batched differently (e.g. if our batch size is 1)? Could we pass through `min_batch_size` and `max_batch_size` to  force the batching to a reasonable size?

Reply:
FWIW, this whole setup may just cause the job to fail and retry a few times (in which case the failed counter would probably be 3 since we'd do that many retries). One testing alternative would be to optionally track if we've failed an inference in the FakeModelHandlerFailsOnInferenceArgs and succeed on retry. That would simulate a transient failure

Reply:
> And it seems like you're saying that if the inference fails, the test will retry the batch a certain number of times. If that's consistent, we could force the batching to make sure we have a reliable count. For the moment I've just changed it to assert that the failed_batches_counter will be greater than zero, but I could make that more precise if you can confirm my interpretation of what you're saying.

I'm not 100% sure what the counter behavior is, but yeah it should retry the batch 3 times; your understanding is correct

> I'm not totally following what you're saying about transient failures though. Do you mean, set up a FakeModelHandler to fail the first time and then succeed, to see if we count the failure?

Yeah

Reply:
I've reduced the examples array to a length of one so hopefully only one batch will run.
Are the retries handled within run_inference()? If so, I expect the error is only thrown up to process() once the retries have been attempted, so I think even if the batch is retried a few times, we'd only increment the counter once, since the incrementation only happens when the error is handled in process(), which should stop all further execution of the method. Let me know if I'm wrong on that.
Also, if that is true and we're only counting failures once, I'm not sure there's use in testing on a transient failure, since the error would never reach process() to trigger the incrementation.

Reply:
We're not currently handling failures in run_inference itself, right now retries will just happen at the pipeline transform level (Beam automatically retries failed bundles). See our conversation here - https://github.com/apache/beam/issues/21437#issuecomment-1263620466 - for more context, and let me know if you want to chat synchronously further here.